// @spud.gg/api/dist/index.js
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {
      get: all[name],
      enumerable: true,
      configurable: true,
      set: (newValue) => all[name] = () => newValue
    });
};

// src/gamepad-util.ts
var exports_gamepad_util = {};
__export(exports_gamepad_util, {
  zeroTinyValues: () => zeroTinyValues,
  wasButtonPressed: () => wasButtonPressed,
  vertical: () => vertical,
  snapToRadian: () => snapToRadian,
  snap8: () => snap8,
  snap4: () => snap4,
  resume: () => resume,
  readAxis: () => readAxis,
  radiansToDegrees: () => radiansToDegrees,
  previousGamepadButtonsDown: () => previousGamepadButtonsDown,
  normalizeAxisValue: () => normalizeAxisValue,
  motion: () => motion,
  magnitude: () => magnitude,
  lastActiveGamepad: () => lastActiveGamepad,
  isButtonPressed: () => isButtonPressed,
  horizontal: () => horizontal,
  getGamepadInfo: () => getGamepadInfo,
  gamepadJustDisconnected: () => gamepadJustDisconnected,
  gamepadJustConnected: () => gamepadJustConnected,
  gamepadButtonJustReleased: () => gamepadButtonJustReleased,
  gamepadButtonJustPressed: () => gamepadButtonJustPressed,
  gamepadButtonHeld: () => gamepadButtonHeld,
  degreesToCardinalDirection: () => degreesToCardinalDirection,
  defaultDeadzone: () => defaultDeadzone,
  clearInputs: () => clearInputs,
  angle: () => angle
});

// src/gamepad-ids.ts
var GamepadBaseLayout = {
  Xbox: "Xbox",
  PlayStation: "PlayStation",
  Nintendo: "Nintendo",
  Generic: "Generic"
};
var GamepadModel = {
  Xbox: "Xbox",
  Xbox360: "Xbox360",
  XboxOne: "XboxOne",
  XboxSX: "XboxSX",
  PlayStation: "PlayStation",
  PlayStation3: "PlayStation3",
  PlayStationDualShock4: "PlayStationDualShock4",
  PlayStationDualShock5: "PlayStationDualShock5",
  PlayStationDualSense: "PlayStationDualSense",
  Nintendo: "Nintendo",
  JoyConL: "JoyConL",
  JoyConR: "JoyConR",
  JoyConPair: "JoyConPair",
  SwitchPro: "SwitchPro",
  Wii: "Wii",
  WiiU: "WiiU",
  SteamController: "SteamController",
  SteamDeck: "SteamDeck",
  Unknown: "Unknown"
};
var { Xbox, PlayStation, Nintendo, Generic } = GamepadBaseLayout;
var gamepadIdentifiers = [
  [1118, 735, "Xbox One S pad", GamepadModel.XboxOne, Xbox],
  [1118, 2821, "Xbox One Elite 2 pad", GamepadModel.XboxOne, Xbox],
  [1118, 2835, "Xbox Series S|X Controller", GamepadModel.XboxSX, Xbox],
  [1118, 2848, "Xbox One S pad", GamepadModel.XboxOne, Xbox],
  [1356, 616, "Sony PS3 Controller", GamepadModel.PlayStation3, PlayStation],
  [1356, 1476, "Sony PS4 DualShock", GamepadModel.PlayStationDualShock4, PlayStation],
  [1356, 2508, "Sony PS4 DualShock", GamepadModel.PlayStationDualShock4, PlayStation],
  [1356, 3302, "Sony PS5 DualSense", GamepadModel.PlayStationDualSense, PlayStation],
  [1356, 2976, "Sony PS4 DualShock", GamepadModel.PlayStationDualShock4, PlayStation],
  [1356, 3570, "Sony PS5 DualSense Edge", GamepadModel.PlayStationDualSense, PlayStation],
  [1118, 765, "Microsoft Xbox One S Controller", GamepadModel.XboxOne, Xbox],
  [1118, 2850, "Xbox Elite Series 2", GamepadModel.XboxSX, Xbox],
  [1118, 514, "Xbox pad v1 (US)", GamepadModel.Xbox, Xbox],
  [1118, 645, "Xbox pad (Japan)", GamepadModel.Xbox, Xbox],
  [1118, 647, "Xbox Controller S", GamepadModel.Xbox, Xbox],
  [1118, 648, "Xbox Controller S v2", GamepadModel.Xbox, Xbox],
  [1118, 649, "Xbox pad v2 (US)", GamepadModel.Xbox, Xbox],
  [1118, 654, "Xbox 360 pad", GamepadModel.Xbox360, Xbox],
  [1118, 655, "Xbox 360 pad v2", GamepadModel.Xbox360, Xbox],
  [1118, 657, "Xbox 360 Wireless Receiver (XBOX)", GamepadModel.Xbox360, Xbox],
  [1118, 681, "Xbox 360 Wireless Receiver (Unofficial)", GamepadModel.Xbox360, Xbox],
  [1118, 721, "Xbox One pad", GamepadModel.XboxOne, Xbox],
  [1118, 733, "Xbox One pad (Firmware 2015)", GamepadModel.XboxOne, Xbox],
  [1118, 739, "Xbox One Elite pad", GamepadModel.XboxOne, Xbox],
  [1118, 746, "Xbox One S pad", GamepadModel.XboxOne, Xbox],
  [1118, 1817, "Xbox 360 Wireless Receiver", GamepadModel.Xbox360, Xbox],
  [1118, 2816, "Xbox One Elite 2 pad", GamepadModel.XboxOne, Xbox],
  [1118, 2826, "Xbox Adaptive Controller", GamepadModel.XboxOne, Xbox],
  [1118, 2834, "Xbox Series S|X Controller", GamepadModel.XboxSX, Xbox],
  [121, 6356, "GPD Win 2 X-Box Controller", GamepadModel.Xbox360, Xbox],
  [1003, 65281, "Wooting One (Legacy)", GamepadModel.Xbox360, Xbox],
  [1003, 65282, "Wooting Two (Legacy)", GamepadModel.Xbox360, Xbox],
  [1008, 909, "HyperX Clutch", GamepadModel.Xbox360, Xbox],
  [1008, 1165, "HyperX Clutch", GamepadModel.Xbox360, Xbox],
  [1008, 1173, "HyperX Clutch Gladiate", GamepadModel.XboxOne, Xbox],
  [1008, 1952, "HyperX Clutch Gladiate RGB", GamepadModel.XboxOne, Xbox],
  [1008, 2230, "HyperX Clutch Gladiate", GamepadModel.XboxOne, Xbox],
  [1008, 2484, "HyperX Clutch Tanto", GamepadModel.XboxOne, Xbox],
  [1103, 3840, "Thrustmaster Wheel", GamepadModel.Xbox, Xbox],
  [1103, 3843, "Thrustmaster Wheel", GamepadModel.Xbox, Xbox],
  [1103, 3847, "Thrustmaster Controller", GamepadModel.Xbox, Xbox],
  [1103, 3856, "Thrustmaster Modena GT Wheel", GamepadModel.Xbox, Xbox],
  [1103, 45862, "Thrustmaster Gamepad GP XID", GamepadModel.Xbox360, Xbox],
  [1103, 53278, "ThrustMaster ESWAP X 2 ELDEN RING EDITION", GamepadModel.XboxOne, Xbox],
  [1133, 49693, "Logitech Gamepad F310", GamepadModel.Xbox360, Xbox],
  [1133, 49694, "Logitech Gamepad F510", GamepadModel.Xbox360, Xbox],
  [1133, 49695, "Logitech Gamepad F710", GamepadModel.Xbox360, Xbox],
  [1133, 49730, "Logitech Chillstream Controller", GamepadModel.Xbox360, Xbox],
  [1133, 51844, "Logitech Xbox Cordless Controller", GamepadModel.Xbox, Xbox],
  [1133, 51848, "Logitech Compact Controller for Xbox", GamepadModel.Xbox, Xbox],
  [1133, 51850, "Logitech Precision Vibration Feedback Wheel", GamepadModel.Xbox, Xbox],
  [1133, 51875, "Logitech DriveFx Racing Wheel", GamepadModel.Xbox360, Xbox],
  [1390, 8196, "Elecom JC-U3613M", GamepadModel.Xbox360, Xbox],
  [1533, 4103, "Mad Catz Controller (unverified)", GamepadModel.Xbox, Xbox],
  [1533, 4218, "InterAct 'PowerPad Pro' X-Box pad (Germany)", GamepadModel.Xbox, Xbox],
  [1534, 12336, "Chic Controller", GamepadModel.Xbox, Xbox],
  [1534, 12337, "Chic Controller", GamepadModel.Xbox, Xbox],
  [1578, 32, "Logic3 Xbox GamePad", GamepadModel.Xbox, Xbox],
  [1578, 51, "Competition Pro Steering Wheel", GamepadModel.Xbox, Xbox],
  [1699, 512, "Saitek Racing Wheel", GamepadModel.Xbox, Xbox],
  [1699, 513, "Saitek Adrenalin", GamepadModel.Xbox, Xbox],
  [1699, 62746, "Saitek P3600", GamepadModel.Xbox360, Xbox],
  [1848, 17667, "Mad Catz Racing Wheel", GamepadModel.XboxOne, Xbox],
  [1848, 17670, "Mad Catz 4506 Wireless Controller", GamepadModel.Xbox, Xbox],
  [1848, 17686, "Mad Catz Control Pad", GamepadModel.Xbox, Xbox],
  [1848, 17696, "Mad Catz Control Pad Pro", GamepadModel.Xbox, Xbox],
  [1848, 17698, "Mad Catz LumiCON", GamepadModel.Xbox, Xbox],
  [1848, 17702, "Mad Catz Control Pad Pro", GamepadModel.Xbox, Xbox],
  [1848, 17712, "Mad Catz Universal MC2 Racing Wheel and Pedals", GamepadModel.Xbox, Xbox],
  [1848, 17718, "Mad Catz MicroCON", GamepadModel.Xbox, Xbox],
  [1848, 17728, "Mad Catz Beat Pad", GamepadModel.Xbox, Xbox],
  [1848, 17750, "Mad Catz Lynx Wireless Controller", GamepadModel.Xbox, Xbox],
  [1848, 17798, "Mad Catz MicroCon Wireless Controller", GamepadModel.Xbox, Xbox],
  [1848, 17800, "Mad Catz Blaster", GamepadModel.Xbox, Xbox],
  [1848, 17919, "Mad Catz Beat Pad (w/ Handle)", GamepadModel.Xbox, Xbox],
  [1848, 18198, "Mad Catz Wired Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [1848, 18200, "Mad Catz Street Fighter IV FightStick SE", GamepadModel.Xbox360, Xbox],
  [1848, 18214, "Mad Catz Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [1848, 18216, "Mad Catz Street Fighter IV FightPad", GamepadModel.Xbox360, Xbox],
  [1848, 18230, "Mad Catz MicroCon Gamepad", GamepadModel.Xbox360, Xbox],
  [1848, 18232, "Mad Catz Wired Xbox 360 Controller (SFIV)", GamepadModel.Xbox360, Xbox],
  [1848, 18240, "Mad Catz Beat Pad", GamepadModel.Xbox360, Xbox],
  [1848, 18243, "Mad Catz Beat Pad Pro", GamepadModel.Xbox, Xbox],
  [1848, 18264, "Mad Catz Arcade Game Stick", GamepadModel.Xbox360, Xbox],
  [1848, 18945, "Mad Catz FightStick TE 2", GamepadModel.XboxOne, Xbox],
  [1848, 24640, "Mad Catz Beat Pad Pro", GamepadModel.Xbox, Xbox],
  [1848, 39025, "Mad Catz Portable Drum", GamepadModel.Xbox360, Xbox],
  [1848, 46886, "Mad Catz Xbox controller - MW2", GamepadModel.Xbox360, Xbox],
  [1848, 46904, "Mad Catz MVC2TE Stick 2", GamepadModel.Xbox360, Xbox],
  [1848, 48879, "Mad Catz JOYTECH NEO SE Advanced GamePad", GamepadModel.Xbox360, Xbox],
  [1848, 51970, "Saitek Cyborg Rumble Pad - PC/Xbox 360", GamepadModel.Xbox360, Xbox],
  [1848, 51971, "Saitek P3200 Rumble Pad - PC/Xbox 360", GamepadModel.Xbox360, Xbox],
  [1848, 52009, "Saitek Aviator Stick AV8R02", GamepadModel.Xbox360, Xbox],
  [1848, 63288, "Super SFIV FightStick TE S", GamepadModel.Xbox360, Xbox],
  [2047, 65535, "Mad Catz GamePad", GamepadModel.Xbox360, Xbox],
  [2821, 6712, "ASUS ROG RAIKIRI", GamepadModel.XboxOne, Xbox],
  [2821, 6843, "ASUS ROG RAIKIRI PRO", GamepadModel.XboxOne, Xbox],
  [3090, 5, "Intec wireless", GamepadModel.Xbox, Xbox],
  [3090, 34817, "Nyko Xbox Controller", GamepadModel.Xbox, Xbox],
  [3090, 34818, "Zeroplus Xbox Controller", GamepadModel.Xbox, Xbox],
  [3090, 34825, "RedOctane Xbox Dance Pad", GamepadModel.Xbox, Xbox],
  [3090, 34826, "Pelican Eclipse PL-2023", GamepadModel.Xbox, Xbox],
  [3090, 34832, "Zeroplus Xbox Controller", GamepadModel.Xbox, Xbox],
  [3090, 39170, "HAMA VibraX - *FAULTY HARDWARE*", GamepadModel.Xbox, Xbox],
  [3375, 2, "Andamiro Pump It Up pad", GamepadModel.Xbox, Xbox],
  [3504, 6401, "Micro Star International Xbox360 Controller for Windows", GamepadModel.Xbox360, Xbox],
  [3660, 4247, "Radica Gamester Controller", GamepadModel.Xbox, Xbox],
  [3660, 4355, "Radica Gamester Reflex", GamepadModel.Xbox, Xbox],
  [3660, 9104, "Radica Games Jtech Controller", GamepadModel.Xbox, Xbox],
  [3660, 13584, "Radica Gamester", GamepadModel.Xbox, Xbox],
  [3695, 3, "Logic3 Freebird wireless Controller", GamepadModel.Xbox, Xbox],
  [3695, 5, "Eclipse wireless Controller", GamepadModel.Xbox, Xbox],
  [3695, 6, "Edge wireless Controller", GamepadModel.Xbox, Xbox],
  [3695, 8, "After Glow Pro Controller", GamepadModel.Xbox, Xbox],
  [3695, 261, "HSM3 Xbox360 dancepad", GamepadModel.Xbox360, Xbox],
  [3695, 275, "Afterglow AX.1 Gamepad for Xbox 360", GamepadModel.Xbox360, Xbox],
  [3695, 287, "Rock Candy Gamepad Wired Controller", GamepadModel.Xbox360, Xbox],
  [3695, 305, "PDP EA Sports Controller", GamepadModel.Xbox360, Xbox],
  [3695, 307, "Xbox 360 Wired Controller", GamepadModel.Xbox360, Xbox],
  [3695, 313, "Afterglow Prismatic Wired Controller", GamepadModel.XboxOne, Xbox],
  [3695, 314, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 326, "Rock Candy Wired Controller for Xbox One", GamepadModel.XboxOne, Xbox],
  [3695, 327, "PDP Marvel Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 348, "PDP Xbox One Arcade Stick", GamepadModel.XboxOne, Xbox],
  [3695, 349, "PDP Mirror's Edge Official Wired Controller for Xbox One", GamepadModel.XboxOne, Xbox],
  [3695, 353, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 354, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 355, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 356, "PDP Battlefield One", GamepadModel.XboxOne, Xbox],
  [3695, 357, "PDP Titanfall 2", GamepadModel.XboxOne, Xbox],
  [3695, 513, "Pelican PL-3601 'TSZ' Wired Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [3695, 531, "Afterglow Gamepad for Xbox 360", GamepadModel.Xbox360, Xbox],
  [3695, 543, "Rock Candy Gamepad for Xbox 360", GamepadModel.Xbox360, Xbox],
  [3695, 582, "Rock Candy Gamepad for Xbox One 2015", GamepadModel.XboxOne, Xbox],
  [3695, 672, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 673, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 674, "PDP Wired Controller for Xbox One - Crimson Red", GamepadModel.XboxOne, Xbox],
  [3695, 676, "PDP Wired Controller for Xbox One - Stealth Series", GamepadModel.XboxOne, Xbox],
  [3695, 678, "PDP Wired Controller for Xbox One - Camo Series", GamepadModel.XboxOne, Xbox],
  [3695, 679, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 680, "PDP Xbox One Controller", GamepadModel.XboxOne, Xbox],
  [3695, 683, "PDP Controller for Xbox One", GamepadModel.XboxOne, Xbox],
  [3695, 685, "PDP Wired Controller for Xbox One - Stealth Series", GamepadModel.XboxOne, Xbox],
  [3695, 691, "Afterglow Prismatic Wired Controller", GamepadModel.XboxOne, Xbox],
  [3695, 696, "Afterglow Prismatic Wired Controller", GamepadModel.XboxOne, Xbox],
  [3695, 769, "Logic3 Controller", GamepadModel.Xbox360, Xbox],
  [3695, 838, "Rock Candy Gamepad for Xbox One 2016", GamepadModel.XboxOne, Xbox],
  [3695, 1025, "Logic3 Controller", GamepadModel.Xbox360, Xbox],
  [3695, 1043, "Afterglow AX.1 Gamepad for Xbox 360", GamepadModel.Xbox360, Xbox],
  [3695, 1281, "PDP Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [3695, 63744, "PDP Afterglow AX.1", GamepadModel.Xbox360, Xbox],
  [3727, 513, "SmartJoy Frag Xpad/PS2 adaptor", GamepadModel.Xbox, Xbox],
  [3727, 12296, "Generic xbox control (dealextreme)", GamepadModel.Xbox, Xbox],
  [3853, 10, "Hori Co. DOA4 FightStick", GamepadModel.Xbox360, Xbox],
  [3853, 12, "Hori PadEX Turbo", GamepadModel.Xbox360, Xbox],
  [3853, 13, "Hori Fighting Stick EX2", GamepadModel.Xbox360, Xbox],
  [3853, 22, "Hori Real Arcade Pro.EX", GamepadModel.Xbox360, Xbox],
  [3853, 27, "Hori Real Arcade Pro VX", GamepadModel.Xbox360, Xbox],
  [3853, 99, "Hori Real Arcade Pro Hayabusa (USA) Xbox One", GamepadModel.XboxOne, Xbox],
  [3853, 103, "HORIPAD ONE", GamepadModel.XboxOne, Xbox],
  [3853, 120, "Hori Real Arcade Pro V Kai Xbox One", GamepadModel.XboxOne, Xbox],
  [3853, 197, "Hori Fighting Commander ONE", GamepadModel.XboxOne, Xbox],
  [3853, 220, "HORIPAD FPS for Nintendo Switch", GamepadModel.Xbox360, Xbox],
  [3853, 337, "Hori Racing Wheel Overdrive for Xbox Series X", GamepadModel.XboxSX, Xbox],
  [3853, 338, "Hori Racing Wheel Overdrive for Xbox Series X", GamepadModel.XboxSX, Xbox],
  [3853, 434, "HORI Taiko No Tatsujin Drum Controller", GamepadModel.XboxOne, Xbox],
  [3888, 267, "Philips Recoil", GamepadModel.Xbox, Xbox],
  [3888, 514, "Joytech Advanced Controller", GamepadModel.Xbox, Xbox],
  [3888, 34952, "BigBen XBMiniPad Controller", GamepadModel.Xbox, Xbox],
  [4140, 65292, "Joytech Wireless Advanced Controller", GamepadModel.Xbox, Xbox],
  [4152, 5168, "SteelSeries Stratus Duo", GamepadModel.Xbox360, Xbox],
  [4152, 5169, "SteelSeries Stratus Duo", GamepadModel.Xbox360, Xbox],
  [4341, 28677, "Turtle Beach Recon Controller", GamepadModel.XboxOne, Xbox],
  [4341, 28680, "Turtle Beach Recon Controller", GamepadModel.XboxOne, Xbox],
  [4341, 28787, "Turtle Beach Stealth Ultra Controller", GamepadModel.XboxOne, Xbox],
  [4553, 22000, "Nacon GC-100XF", GamepadModel.Xbox360, Xbox],
  [4607, 1297, "PXN V900", GamepadModel.Xbox360, Xbox],
  [4617, 10370, "Ardwiino Controller", GamepadModel.Xbox360, Xbox],
  [4779, 4, "Honey Bee Xbox360 dancepad", GamepadModel.Xbox360, Xbox],
  [4779, 769, "PDP AFTERGLOW AX.1", GamepadModel.Xbox360, Xbox],
  [4779, 771, "Mortal Kombat Klassic FightStick", GamepadModel.Xbox360, Xbox],
  [4779, 34825, "Xbox DDR dancepad", GamepadModel.Xbox, Xbox],
  [5168, 18248, "RedOctane Guitar Hero X-plorer", GamepadModel.Xbox360, Xbox],
  [5168, 34952, "TX6500+ Dance Pad (first generation)", GamepadModel.Xbox, Xbox],
  [5168, 63489, "RedOctane Controller", GamepadModel.Xbox360, Xbox],
  [5227, 1537, "BigBen Interactive XBOX 360 Controller", GamepadModel.Xbox360, Xbox],
  [5227, 1540, "Bigben Interactive DAIJA Arcade Stick", GamepadModel.Xbox360, Xbox],
  [5426, 2560, "Razer Atrox Arcade Stick", GamepadModel.XboxOne, Xbox],
  [5426, 2563, "Razer Wildcat", GamepadModel.XboxOne, Xbox],
  [5426, 2601, "Razer Wolverine V2", GamepadModel.XboxOne, Xbox],
  [5604, 16128, "Power A Mini Pro Elite", GamepadModel.Xbox360, Xbox],
  [5604, 16138, "Xbox Airflo wired controller", GamepadModel.Xbox360, Xbox],
  [5604, 16144, "Batarang Xbox 360 controller", GamepadModel.Xbox360, Xbox],
  [5678, 48879, "Joytech Neo-Se Take2", GamepadModel.Xbox360, Xbox],
  [5769, 64768, "Razer Onza Tournament Edition", GamepadModel.Xbox360, Xbox],
  [5769, 64769, "Razer Onza Classic Edition", GamepadModel.Xbox360, Xbox],
  [5769, 65024, "Razer Sabertooth", GamepadModel.Xbox360, Xbox],
  [6127, 24962, "Lenovo Legion Controller for Windows", GamepadModel.Xbox360, Xbox],
  [6473, 1050, "Amazon Game Controller", GamepadModel.Xbox360, Xbox],
  [6790, 58128, "Legion Go S", GamepadModel.Xbox360, Xbox],
  [7085, 2, "Harmonix Rock Band Guitar", GamepadModel.Xbox360, Xbox],
  [7085, 3, "Harmonix Rock Band Drumkit", GamepadModel.Xbox360, Xbox],
  [7085, 304, "Ion Drum Rocker", GamepadModel.Xbox360, Xbox],
  [7085, 61462, "Mad Catz Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [7085, 61464, "Mad Catz Street Fighter IV SE Fighting Stick", GamepadModel.Xbox360, Xbox],
  [7085, 61465, "Mad Catz Brawlstick for Xbox 360", GamepadModel.Xbox360, Xbox],
  [7085, 61473, "Mad Cats Ghost Recon FS GamePad", GamepadModel.Xbox360, Xbox],
  [7085, 61475, "MLG Pro Circuit Controller (Xbox)", GamepadModel.Xbox360, Xbox],
  [7085, 61477, "Mad Catz Call Of Duty", GamepadModel.Xbox360, Xbox],
  [7085, 61479, "Mad Catz FPS Pro", GamepadModel.Xbox360, Xbox],
  [7085, 61480, "Street Fighter IV FightPad", GamepadModel.Xbox360, Xbox],
  [7085, 61486, "Mad Catz Fightpad", GamepadModel.Xbox360, Xbox],
  [7085, 61488, "Mad Catz Xbox 360 MC2 MicroCon Racing Wheel", GamepadModel.Xbox360, Xbox],
  [7085, 61494, "Mad Catz MicroCon GamePad Pro", GamepadModel.Xbox360, Xbox],
  [7085, 61496, "Street Fighter IV FightStick TE", GamepadModel.Xbox360, Xbox],
  [7085, 61497, "Mad Catz MvC2 TE", GamepadModel.Xbox360, Xbox],
  [7085, 61498, "Mad Catz SFxT Fightstick Pro", GamepadModel.Xbox360, Xbox],
  [7085, 61501, "Street Fighter IV Arcade Stick TE - Chun Li", GamepadModel.Xbox360, Xbox],
  [7085, 61502, "Mad Catz MLG FightStick TE", GamepadModel.Xbox360, Xbox],
  [7085, 61503, "Mad Catz FightStick SoulCaliber", GamepadModel.Xbox360, Xbox],
  [7085, 61506, "Mad Catz FightStick TES+", GamepadModel.Xbox360, Xbox],
  [7085, 61568, "Mad Catz FightStick TE2", GamepadModel.Xbox360, Xbox],
  [7085, 62721, "HoriPad EX2 Turbo", GamepadModel.Xbox360, Xbox],
  [7085, 62722, "Hori Real Arcade Pro.VX SA", GamepadModel.Xbox360, Xbox],
  [7085, 62723, "Hori Fighting Stick VX", GamepadModel.Xbox360, Xbox],
  [7085, 62724, "Hori Real Arcade Pro. EX", GamepadModel.Xbox360, Xbox],
  [7085, 62725, "Hori Fighting Stick EX2B", GamepadModel.Xbox360, Xbox],
  [7085, 62726, "Hori Real Arcade Pro.EX Premium VLX", GamepadModel.Xbox360, Xbox],
  [7085, 63744, "Harmonix Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [7085, 63745, "Gamestop Xbox 360 Controller", GamepadModel.Xbox360, Xbox],
  [7085, 63747, "Tron Xbox 360 controller", GamepadModel.Xbox360, Xbox],
  [7085, 63748, "PDP Versus Fighting Pad", GamepadModel.Xbox360, Xbox],
  [7085, 63750, "MortalKombat FightStick", GamepadModel.Xbox360, Xbox],
  [7085, 64001, "MadCatz GamePad", GamepadModel.Xbox360, Xbox],
  [7085, 64768, "Razer Onza TE", GamepadModel.Xbox360, Xbox],
  [7085, 64769, "Razer Onza", GamepadModel.Xbox360, Xbox],
  [7913, 5520, "ZOTAC Gaming Zone", GamepadModel.Xbox360, Xbox],
  [8406, 8193, "BDA Xbox Series X Wired Controller", GamepadModel.XboxSX, Xbox],
  [8406, 8201, "PowerA Enhanced Wired Controller for Xbox Series X|S", GamepadModel.XboxSX, Xbox],
  [8406, 8292, "PowerA Wired Controller for Xbox", GamepadModel.XboxOne, Xbox],
  [8406, 10271, "PowerA Wired Controller For Xbox 360", GamepadModel.Xbox360, Xbox],
  [8406, 16395, "PowerA FUSION Pro 4 Wired Controller", GamepadModel.XboxOne, Xbox],
  [8406, 35083, "PowerA MOGA XP-Ultra Controller", GamepadModel.XboxOne, Xbox],
  [9029, 57355, "Machenike G5 Pro Controller", GamepadModel.Xbox360, Xbox],
  [9414, 20480, "Razer Atrox Arcade Stick", GamepadModel.Xbox360, Xbox],
  [9414, 21248, "PowerA MINI PROEX Controller", GamepadModel.Xbox360, Xbox],
  [9414, 21251, "Xbox Airflo wired controller", GamepadModel.Xbox360, Xbox],
  [9414, 21258, "Xbox 360 Pro EX Controller", GamepadModel.Xbox360, Xbox],
  [9414, 21274, "PowerA Pro Ex", GamepadModel.Xbox360, Xbox],
  [9414, 21399, "FUS1ON Tournament Controller", GamepadModel.Xbox360, Xbox],
  [9414, 21530, "PowerA Xbox One Mini Wired Controller", GamepadModel.XboxOne, Xbox],
  [9414, 21546, "Xbox ONE spectra", GamepadModel.XboxOne, Xbox],
  [9414, 21562, "PowerA Xbox One wired controller", GamepadModel.XboxOne, Xbox],
  [9414, 21760, "Hori XBOX 360 EX 2 with Turbo", GamepadModel.Xbox360, Xbox],
  [9414, 21761, "Hori Real Arcade Pro VX-SA", GamepadModel.Xbox360, Xbox],
  [9414, 21762, "Hori Fighting Stick VX Alt", GamepadModel.Xbox360, Xbox],
  [9414, 21763, "Hori Fighting Edge", GamepadModel.Xbox360, Xbox],
  [9414, 21766, "Hori SOULCALIBUR V Stick", GamepadModel.Xbox360, Xbox],
  [9414, 21773, "Hori GEM Xbox controller", GamepadModel.Xbox360, Xbox],
  [9414, 21774, "Hori Real Arcade Pro V Kai 360", GamepadModel.Xbox360, Xbox],
  [9414, 21776, "Hori Fighting Commander ONE (Xbox 360/PC Mode)", GamepadModel.Xbox360, Xbox],
  [9414, 21786, "PowerA FUSION Pro Controller", GamepadModel.XboxOne, Xbox],
  [9414, 22042, "PowerA FUSION Controller", GamepadModel.XboxOne, Xbox],
  [9414, 22554, "ThrustMaster XB1 Classic Controller", GamepadModel.XboxOne, Xbox],
  [9414, 23296, "ThrustMaster Ferrari 458 Racing Wheel", GamepadModel.Xbox360, Xbox],
  [9414, 23298, "Thrustmaster, Inc. GPX Controller", GamepadModel.Xbox360, Xbox],
  [9414, 23299, "Thrustmaster Ferrari 458 Racing Wheel", GamepadModel.Xbox360, Xbox],
  [9414, 23812, "Razer Sabertooth", GamepadModel.Xbox360, Xbox],
  [9414, 64254, "Rock Candy Gamepad for Xbox 360", GamepadModel.Xbox360, Xbox],
  [9571, 1421, "OneXPlayer Gamepad", GamepadModel.Xbox360, Xbox],
  [10571, 13059, "Snakebyte GAMEPAD BASE X", GamepadModel.XboxOne, Xbox],
  [10571, 13316, "Snakebyte GAMEPAD RGB X", GamepadModel.XboxOne, Xbox],
  [10643, 8193, "TECNO Pocket Go", GamepadModel.Xbox360, Xbox],
  [11720, 8192, "8BitDo Pro 2 Wired Controller fox Xbox", GamepadModel.XboxOne, Xbox],
  [11720, 8207, "8BitDo Ultimate 3-mode Controller for Xbox", GamepadModel.XboxOne, Xbox],
  [11720, 12550, "8BitDo Ultimate Wireless / Pro 2 Wired Controller", GamepadModel.Xbox360, Xbox],
  [11720, 12553, "8BitDo Ultimate Wireless Bluetooth", GamepadModel.Xbox360, Xbox],
  [11720, 12554, "8BitDo Ultimate 2C Wireless Controller", GamepadModel.Xbox360, Xbox],
  [11720, 12555, "8BitDo Ultimate 2 Wireless Controller", GamepadModel.Xbox360, Xbox],
  [11720, 24577, "8BitDo SN30 Pro", GamepadModel.Xbox360, Xbox],
  [11812, 1059, "Hyperkin DuchesS Xbox One pad", GamepadModel.XboxOne, Xbox],
  [11812, 1618, "Hyperkin Duke X-Box One pad", GamepadModel.XboxOne, Xbox],
  [11812, 5768, "Hyperkin X91 X-Box One pad", GamepadModel.XboxOne, Xbox],
  [11925, 1284, "SCUF Gaming Controller", GamepadModel.XboxOne, Xbox],
  [12771, 4352, "Wooting One", GamepadModel.Xbox360, Xbox],
  [12771, 4608, "Wooting Two", GamepadModel.Xbox360, Xbox],
  [12771, 4624, "Wooting Lekker", GamepadModel.Xbox360, Xbox],
  [12771, 4640, "Wooting Two HE", GamepadModel.Xbox360, Xbox],
  [12771, 4656, "Wooting Two HE (ARM)", GamepadModel.Xbox360, Xbox],
  [12771, 4864, "Wooting 60HE (AVR)", GamepadModel.Xbox360, Xbox],
  [12771, 4880, "Wooting 60HE (ARM)", GamepadModel.Xbox360, Xbox],
  [12933, 1539, "Nacon Pro Compact controller for Xbox", GamepadModel.XboxOne, Xbox],
  [12933, 1543, "Nacon GC-100", GamepadModel.Xbox360, Xbox],
  [12933, 1556, "Nacon Pro Compact", GamepadModel.XboxOne, Xbox],
  [12933, 1606, "Nacon Pro Compact", GamepadModel.XboxOne, Xbox],
  [12933, 1634, "Nacon Revolution5 Pro", GamepadModel.Xbox360, Xbox],
  [12933, 1635, "Nacon Evol-X", GamepadModel.XboxOne, Xbox],
  [13623, 4100, "GameSir T4 Kaleid", GamepadModel.Xbox360, Xbox],
  [13623, 4112, "GameSir G7 SE", GamepadModel.XboxOne, Xbox],
  [13932, 5, "ByoWave Proteus Controller", GamepadModel.XboxOne, Xbox],
  [14183, 257, "Fanatec Speedster 3 Forceshock Wheel", GamepadModel.Xbox, Xbox],
  [16701, 8452, "Black Shark Green Ghost Gamepad", GamepadModel.Xbox360, Xbox],
  [1406, 8198, "Nintendo Joy-Con Left", GamepadModel.JoyConL, Nintendo],
  [1406, 8199, "Nintendo Joy-Con Right", GamepadModel.JoyConR, Nintendo],
  [1406, 8206, "Nintendo Switch Joy-Con Pair", GamepadModel.JoyConPair, Nintendo],
  [1406, 8201, "Nintendo Switch Pro Controller", GamepadModel.SwitchPro, Nintendo],
  [1406, 774, "Wii Remote Controller RVL-003", GamepadModel.Wii, Nintendo],
  [1406, 823, "Wii U GameCube Controller Adapter", GamepadModel.WiiU, Nintendo],
  [1356, 3290, "PlayStation Classic controller", GamepadModel.PlayStation, PlayStation],
  [1356, 981, "PlayStation Move motion controller", GamepadModel.PlayStation, PlayStation],
  [1356, 1071, "PlayStation Move navigation controller", GamepadModel.PlayStation, PlayStation],
  [1118, 7, "Microsoft SideWinder Game Pad", GamepadModel.Unknown, Xbox],
  [1118, 8, "Microsoft SideWinder Precision Pro", GamepadModel.Unknown, Xbox],
  [1118, 14, "Microsoft SideWinder Freestyle Pro", GamepadModel.Unknown, Xbox],
  [1118, 26, "Microsoft SideWinder Precision Racing Wheel", GamepadModel.Unknown, Xbox],
  [1118, 27, "Microsoft SideWinder Force Feedback 2 Joystick", GamepadModel.Unknown, Xbox],
  [1118, 38, "Microsoft SideWinder GamePad Pro", GamepadModel.Unknown, Xbox],
  [1118, 39, "Microsoft SideWinder PnP GamePad", GamepadModel.Unknown, Xbox],
  [1118, 40, "Microsoft SideWinder Dual Strike", GamepadModel.Unknown, Xbox],
  [1118, 51, "Microsoft SideWinder Strategic Commander", GamepadModel.Unknown, Xbox],
  [1118, 52, "Microsoft SideWinder Force Feedback Wheel", GamepadModel.Unknown, Xbox],
  [1118, 56, "Microsoft SideWinder Precision 2", GamepadModel.Unknown, Xbox],
  [1118, 59, "Microsoft SideWinder Game Voice", GamepadModel.Unknown, Xbox],
  [1118, 60, "Microsoft SideWinder Joystick", GamepadModel.Unknown, Xbox],
  [10462, 4354, "Steam Controller", GamepadModel.SteamController, Generic],
  [10462, 4609, "Steam Controller", GamepadModel.SteamController, Generic],
  [10462, 4418, "Wireless Steam Controller", GamepadModel.SteamController, Generic],
  [10462, 4613, "Steam Deck", GamepadModel.SteamDeck, Generic],
  [1118, 673, "Xbox Controller", GamepadModel.Xbox360, Xbox],
  [1133, 49686, "Logitech Dual Action Gamepad F310", GamepadModel.Unknown, Generic],
  [1133, 49688, "Logitech RumblePad F510", GamepadModel.Unknown, Generic],
  [1133, 49689, "Logitech RumblePad F710", GamepadModel.Unknown, Generic],
  [1133, 49690, "Logitech Precision Gamepad", GamepadModel.Unknown, Generic],
  [2064, 1, "PCS USB Gamepad", GamepadModel.Unknown, Generic],
  [2064, 3, "PCS PS3/PC Gamepad", GamepadModel.PlayStation3, PlayStation],
  [2064, 3, "PCS PlayStation Gamepad", GamepadModel.PlayStation, PlayStation],
  [2064, 57345, "PCS Twin controller", GamepadModel.Unknown, Generic],
  [2064, 58625, "PCS SNES Gamepad", GamepadModel.Unknown, Nintendo],
  [3853, 34, "Hori Real Arcade Pro 3 SA", GamepadModel.Unknown, Generic],
  [3853, 17, "Hori Real Arcade Pro 3", GamepadModel.Unknown, Generic],
  [3853, 144, "Hori Horipad Ultimate", GamepadModel.Unknown, Generic],
  [3853, 193, "Hori HORIPAD for Nintendo Switch", GamepadModel.Unknown, Nintendo],
  [3853, 85, "Hori Fighting Commander 4", GamepadModel.Unknown, GamepadBaseLayout.PlayStation],
  [1103, 46675, "RGT Force Feedback Clutch Racing Wheel", GamepadModel.Unknown, Generic],
  [2389, 29200, "NVIDIA Shield Controller", GamepadModel.Unknown, Xbox],
  [2389, 29204, "NVIDIA Shield Controller (2017)", GamepadModel.Unknown, Xbox]
];
function hexToString(hex) {
  return hex.toString(16).padStart(4, "0");
}
var microsoftVendorId = "045e";
var sonyVendorId = "054c";
var nintendoVendorId = "057e";
function identifyGamepadUnoptimized(rawId) {
  const id = rawId.toLowerCase();
  for (const [rawVendorId, rawProductId, productName, generalModel, layout] of gamepadIdentifiers) {
    const vendorId = hexToString(rawVendorId);
    const productId = hexToString(rawProductId);
    if (id.includes(vendorId) && id.includes(productId)) {
      return { layout, generalModel, vendorId, productId, productName };
    }
  }
  if (id.includes(microsoftVendorId)) {
    return {
      layout: GamepadBaseLayout.Xbox,
      generalModel: GamepadModel.Xbox,
      vendorId: microsoftVendorId,
      productId: "0000",
      productName: "Unknown"
    };
  } else if (id.includes(sonyVendorId)) {
    return {
      layout: GamepadBaseLayout.PlayStation,
      generalModel: GamepadModel.PlayStation,
      vendorId: sonyVendorId,
      productId: "0000",
      productName: "Unknown"
    };
  } else if (id.includes(nintendoVendorId)) {
    return {
      layout: GamepadBaseLayout.Nintendo,
      generalModel: GamepadModel.Nintendo,
      vendorId: nintendoVendorId,
      productId: "0000",
      productName: "Unknown"
    };
  }
  if (id.includes("xbox")) {
    return {
      layout: GamepadBaseLayout.Xbox,
      generalModel: GamepadModel.Xbox,
      vendorId: "0000",
      productId: "0000",
      productName: "Unknown"
    };
  } else if (id.includes("playstation") || id.includes("dualsense") || id.includes("dualshock") || id.includes("ps5") || id.includes("ps4")) {
    return {
      layout: GamepadBaseLayout.PlayStation,
      generalModel: GamepadModel.PlayStation,
      vendorId: "0000",
      productId: "0000",
      productName: "Unknown"
    };
  } else if (id.includes("nintendo") || id.includes("joy-con")) {
    return {
      layout: GamepadBaseLayout.Nintendo,
      generalModel: GamepadModel.Nintendo,
      vendorId: "0000",
      productId: "0000",
      productName: "Unknown"
    };
  }
  return {
    layout: GamepadBaseLayout.Generic,
    generalModel: GamepadModel.Unknown,
    vendorId: "0000",
    productId: "0000",
    productName: "Unknown"
  };
}
function memoizeStringFunction(fn) {
  const cache = new Map;
  return (input) => {
    const value = cache.get(input);
    if (value !== undefined)
      return value;
    const result = fn(input);
    cache.set(input, result);
    return result;
  };
}
var identifyGamepad = memoizeStringFunction(identifyGamepadUnoptimized);

// src/util.ts
function assert(condition, message = "Assertion failed") {
  if (!condition)
    throw new Error(message);
}
function includes(array, value) {
  return array.includes(value);
}
function unique(array) {
  return array.filter((value, index, self) => self.indexOf(value) === index);
}
function exists(value) {
  return value !== null && value !== undefined;
}
function clamp(value, min, max) {
  return Math.min(Math.max(value, min), max);
}

// src/types.ts
var Button = {
  DpadUp: 12,
  DpadDown: 13,
  DpadLeft: 14,
  DpadRight: 15,
  Start: 9,
  Select: 8,
  LeftStick: 10,
  RightStick: 11,
  LeftShoulder: 4,
  RightShoulder: 5,
  LeftTrigger: 6,
  RightTrigger: 7,
  South: 0,
  East: 1,
  West: 2,
  North: 3,
  Home: 16,
  Capture: 17,
  TouchPad: 17
};
var buttons = unique(Object.values(Button));
var Axis = {
  LeftStickX: 0,
  LeftStickY: 1,
  RightStickX: 2,
  RightStickY: 3,
  LeftTrigger: 6,
  RightTrigger: 7
};
var axes = Object.values(Axis);
var Side = {
  Left: "Left",
  Right: "Right"
};
var Player = {
  P1: 0,
  P2: 1,
  P3: 2,
  P4: 3
};
var PlayerProperty = {
  p1: "p1",
  p2: "p2",
  p3: "p3",
  p4: "p4"
};
var playerIndexes = [Player.P1, Player.P2, Player.P3, Player.P4];
var playerProperties = [PlayerProperty.p1, PlayerProperty.p2, PlayerProperty.p3, PlayerProperty.p4];
var defaultPlayerNames = ["P1", "P2", "P3", "P4"];
var AnyPlayer = "AnyPlayer";
var AllPlayers = "AllPlayers";
var HapticIntensity = {
  Heavy: "Heavy",
  Balanced: "Balanced",
  Light: "Light"
};
var CardinalDirection = {
  E: "E",
  SE: "SE",
  S: "S",
  SW: "SW",
  W: "W",
  NW: "NW",
  N: "N",
  NE: "NE"
};
var Motion = {
  Up: "Up",
  Down: "Down",
  Left: "Left",
  Right: "Right"
};
var Neutral = "Neutral";

// src/gamepad-util.ts
var previousGamepadButtonsDown = [
  new Set,
  new Set,
  new Set,
  new Set
];
var previousGamepads = [
  null,
  null,
  null,
  null
];
var currentJoystickMotions4Way = new Map;
var previousJoystickMotions4Way = new Map;
var currentJoystickMotions8Way = new Map;
var previousJoystickMotions8Way = new Map;
var resume = {
  time: 0,
  player: AnyPlayer
};
var lastActiveGamepad = 0;
function clearInputs() {
  const gamepads = navigator.getGamepads();
  previousGamepads[0] = gamepads[0] ?? null;
  previousGamepads[1] = gamepads[1] ?? null;
  previousGamepads[2] = gamepads[2] ?? null;
  previousGamepads[3] = gamepads[3] ?? null;
  navigator.getGamepads().forEach((gamepad, i) => {
    if (!gamepad || !gamepad.connected)
      return;
    if (i !== 0 && i !== 1 && i !== 2 && i !== 3)
      return;
    const buttonsDown = previousGamepadButtonsDown[i];
    buttonsDown.clear();
    gamepad.buttons.forEach((button, buttonIndex) => {
      if (button.pressed) {
        buttonsDown.add(buttonIndex);
      }
    });
    const activityThreshold = 0.75;
    if (gamepad.buttons.some((button) => button.pressed)) {
      lastActiveGamepad = i;
    } else {
      const leftStickMagnitude = magnitude(normalizeAxisValue(gamepad.axes[Axis.LeftStickX] ?? 0, defaultDeadzone), normalizeAxisValue(gamepad.axes[Axis.LeftStickY] ?? 0, defaultDeadzone));
      const rightStickMagnitude = magnitude(normalizeAxisValue(gamepad.axes[Axis.RightStickX] ?? 0, defaultDeadzone), normalizeAxisValue(gamepad.axes[Axis.RightStickY] ?? 0, defaultDeadzone));
      if (leftStickMagnitude > activityThreshold || rightStickMagnitude > activityThreshold) {
        lastActiveGamepad = i;
      }
    }
  });
  previousJoystickMotions4Way = currentJoystickMotions4Way;
  currentJoystickMotions4Way = new Map;
  previousJoystickMotions8Way = currentJoystickMotions8Way;
  currentJoystickMotions8Way = new Map;
}
function isButtonPressed(button, player = AnyPlayer) {
  const gamepads = navigator.getGamepads();
  switch (player) {
    case AnyPlayer: {
      return gamepads.some((gamepad) => gamepad?.buttons[button]?.pressed);
    }
    case AllPlayers: {
      const connected = gamepads.filter(exists);
      return connected.length > 0 && connected.every((gamepad) => gamepad.buttons[button]?.pressed);
    }
    default: {
      const gamepad = gamepads[player];
      if (!gamepad)
        return false;
      return !!gamepad.buttons[button]?.pressed;
    }
  }
}
function wasButtonPressed(button, player = AnyPlayer) {
  switch (player) {
    case AnyPlayer: {
      return previousGamepadButtonsDown.some((buttonsDown) => buttonsDown.has(button));
    }
    case AllPlayers: {
      return previousGamepadButtonsDown.every((buttonsDown) => buttonsDown.has(button));
    }
    default: {
      return previousGamepadButtonsDown[player].has(button);
    }
  }
}
function gamepadButtonJustPressed(button, player = AnyPlayer) {
  return isButtonPressed(button, player) && !wasButtonPressed(button, player);
}
function gamepadButtonJustReleased(button, player = AnyPlayer) {
  return !isButtonPressed(button, player) && wasButtonPressed(button, player);
}
function gamepadButtonHeld(button, player = AnyPlayer) {
  return isButtonPressed(button, player) && wasButtonPressed(button, player);
}
function gamepadJustConnected(player) {
  const gamepads = navigator.getGamepads();
  return !!(gamepads[player] && !previousGamepads[player]);
}
function gamepadJustDisconnected(player) {
  const previous = previousGamepads[player];
  if (!previous)
    return false;
  const gamepads = navigator.getGamepads();
  const stillConnected = gamepads[player];
  return !stillConnected;
}
function readAxis(axis, player) {
  const gamepads = navigator.getGamepads();
  if (axis === Axis.LeftTrigger || axis === Axis.RightTrigger) {
    return gamepads[player]?.buttons?.[axis]?.value;
  }
  return gamepads[player]?.axes[axis];
}
function getGamepadInfo(player) {
  const gamepad = navigator.getGamepads()[player];
  if (!gamepad) {
    return {
      id: "",
      index: -1,
      mapping: "standard",
      timestamp: 0,
      baseLayout: GamepadBaseLayout.Generic,
      model: GamepadModel.Unknown,
      vendorId: "0000",
      productId: "0000",
      productName: "Unknown"
    };
  }
  const id = gamepad.id;
  const { index, mapping, timestamp } = gamepad;
  const { generalModel: model, layout: baseLayout, productId, productName, vendorId } = identifyGamepad(id);
  return { id, index, mapping, timestamp, model, vendorId, productId, productName, baseLayout };
}
var defaultDeadzone = 0.25;
var zeroTinyValues = (value) => {
  const epsilon = 0.000001;
  if (Math.abs(value) < epsilon)
    return 0;
  return value;
};
function normalizeAxisValue(value, deadzone = defaultDeadzone) {
  if (value === 0)
    return value;
  if (Math.abs(value) <= deadzone)
    return 0;
  const normalized = (Math.abs(value) - deadzone) / (1 - deadzone);
  const result = Math.sign(value) * normalized;
  return zeroTinyValues(result);
}
function snapToRadian(x, y, numberOfSnaps, deadzone = defaultDeadzone) {
  const angle = Math.atan2(y, x);
  const snapRadians = Math.PI / numberOfSnaps;
  const newAngle = snapRadians * Math.round(angle / snapRadians);
  const magnitude = Math.min(1, Math.sqrt(x * x + y * y));
  if (magnitude <= deadzone) {
    return { x: 0, y: 0 };
  }
  return {
    x: zeroTinyValues(Math.cos(newAngle) * normalizeAxisValue(magnitude, deadzone)),
    y: zeroTinyValues(Math.sin(newAngle) * normalizeAxisValue(magnitude, deadzone))
  };
}
function snap8(x, y, deadzone = defaultDeadzone) {
  return snapToRadian(x, y, 4, deadzone);
}
function snap4(x, y, deadzone = defaultDeadzone) {
  return snapToRadian(x, y, 2, deadzone);
}
function vertical(x, y, deadzone = defaultDeadzone) {
  return {
    x: 0,
    y: snapToRadian(x, y, 2, deadzone).y
  };
}
function horizontal(x, y, deadzone = defaultDeadzone) {
  return {
    x: snapToRadian(x, y, 2, deadzone).x,
    y: 0
  };
}
function magnitude(x, y) {
  return Math.sqrt(x ** 2 + y ** 2);
}
function angle(x, y) {
  return Math.atan2(y, x);
}
function radiansToDegrees(radians) {
  return (radians * 180 / Math.PI + 360) % 360;
}
function degreesToCardinalDirection(angleDegrees) {
  const directions = ["E", "SE", "S", "SW", "W", "NW", "N", "NE"];
  const sectorSize = 360 / directions.length;
  const index = Math.round(angleDegrees / sectorSize) % directions.length;
  return directions[index] ?? "E";
}
function motion({
  rawX,
  rawY,
  deadzone,
  player,
  stick,
  activationThreshold,
  deactivationThreshold,
  diagonals
}) {
  const key = `${player}-${stick}-${activationThreshold}-${deactivationThreshold}`;
  const x = normalizeAxisValue(rawX, deadzone);
  const y = normalizeAxisValue(rawY, deadzone);
  if (diagonals) {
    const prevState2 = previousJoystickMotions8Way.get(key) || Neutral;
    const m = magnitude(x, y);
    const currentState2 = m < deactivationThreshold ? Neutral : m > activationThreshold ? degreesToCardinalDirection(radiansToDegrees(angle(x, y))) : prevState2;
    currentJoystickMotions8Way.set(key, currentState2);
    return {
      justMoved: {
        [CardinalDirection.E]: prevState2 !== currentState2 && currentState2 === CardinalDirection.E,
        [CardinalDirection.SE]: prevState2 !== currentState2 && currentState2 === CardinalDirection.SE,
        [CardinalDirection.S]: prevState2 !== currentState2 && currentState2 === CardinalDirection.S,
        [CardinalDirection.SW]: prevState2 !== currentState2 && currentState2 === CardinalDirection.SW,
        [CardinalDirection.W]: prevState2 !== currentState2 && currentState2 === CardinalDirection.W,
        [CardinalDirection.NW]: prevState2 !== currentState2 && currentState2 === CardinalDirection.NW,
        [CardinalDirection.N]: prevState2 !== currentState2 && currentState2 === CardinalDirection.N,
        [CardinalDirection.NE]: prevState2 !== currentState2 && currentState2 === CardinalDirection.NE
      },
      direction: currentState2,
      get didJustMove() {
        return this.justMoved[CardinalDirection.E] || this.justMoved[CardinalDirection.SE] || this.justMoved[CardinalDirection.S] || this.justMoved[CardinalDirection.SW] || this.justMoved[CardinalDirection.W] || this.justMoved[CardinalDirection.NW] || this.justMoved[CardinalDirection.N] || this.justMoved[CardinalDirection.NE];
      }
    };
  }
  const prevState = previousJoystickMotions4Way.get(key) || { x: Neutral, y: Neutral };
  const currentState = structuredClone(prevState);
  if (Math.abs(x) < deactivationThreshold) {
    currentState.x = Neutral;
  } else if (x > activationThreshold) {
    currentState.x = Motion.Right;
  } else if (x < -activationThreshold) {
    currentState.x = Motion.Left;
  }
  if (Math.abs(y) < deactivationThreshold) {
    currentState.y = Neutral;
  } else if (y > activationThreshold) {
    currentState.y = Motion.Down;
  } else if (y < -activationThreshold) {
    currentState.y = Motion.Up;
  }
  currentJoystickMotions4Way.set(key, currentState);
  return {
    justMoved: {
      [Motion.Up]: prevState.y !== Motion.Up && currentState.y === Motion.Up,
      [Motion.Down]: prevState.y !== Motion.Down && currentState.y === Motion.Down,
      [Motion.Left]: prevState.x !== Motion.Left && currentState.x === Motion.Left,
      [Motion.Right]: prevState.x !== Motion.Right && currentState.x === Motion.Right
    },
    direction: currentState,
    get didJustMove() {
      return this.justMoved[Motion.Up] || this.justMoved[Motion.Down] || this.justMoved[Motion.Left] || this.justMoved[Motion.Right];
    }
  };
}

// src/haptics.ts
var mapHapticIntensityToTriggerStrength = {
  [HapticIntensity.Light]: 0.05,
  [HapticIntensity.Balanced]: 0.3,
  [HapticIntensity.Heavy]: 1
};
var warnings = {
  [Player.P1]: false,
  [Player.P2]: false,
  [Player.P3]: false,
  [Player.P4]: false
};
var warnDisconnected = (player) => {
  if (warnings[player])
    return;
  warnings[player] = true;
  console.warn(`Cannot update haptic for player ${player + 1}. Controller not connected!`);
};
var warnUnsupported = (player) => {
  if (warnings[player])
    return;
  warnings[player] = true;
  console.warn(`Haptics not supported for player ${player + 1}.`);
};
async function haptic({
  player,
  duration = 0,
  strongMagnitude = 0,
  weakMagnitude = 0
}) {
  const gamepad = navigator.getGamepads()[player];
  if (!gamepad || !gamepad.connected)
    return warnDisconnected(player);
  if (!gamepad.vibrationActuator)
    return warnUnsupported(player);
  return await gamepad.vibrationActuator.playEffect("dual-rumble", {
    duration,
    weakMagnitude,
    strongMagnitude
  });
}
async function leftTriggerHaptic({
  player,
  duration = 0,
  intensity = 0
}) {
  const gamepad = navigator.getGamepads()[player];
  if (!gamepad || !gamepad.connected)
    return warnDisconnected(player);
  if (!gamepad.vibrationActuator)
    return warnUnsupported(player);
  return await gamepad.vibrationActuator.playEffect("trigger-rumble", {
    duration,
    leftTrigger: intensity
  });
}
async function rightTriggerHaptic({
  player,
  duration = 0,
  intensity = 0
}) {
  const gamepad = navigator.getGamepads()[player];
  if (!gamepad || !gamepad.connected)
    return warnDisconnected(player);
  if (!gamepad.vibrationActuator)
    return warnUnsupported(player);
  return await gamepad.vibrationActuator.playEffect("trigger-rumble", {
    duration,
    rightTrigger: intensity
  });
}
function makePlayerHaptic(player) {
  return async function haptic2({
    duration = 0,
    strong = 0,
    weak = 0
  }) {
    const gamepad = navigator.getGamepads()[player];
    if (!gamepad || !gamepad.connected)
      return warnDisconnected(player);
    if (!gamepad.vibrationActuator)
      return warnUnsupported(player);
    return await gamepad.vibrationActuator.playEffect("dual-rumble", {
      duration,
      weakMagnitude: weak,
      strongMagnitude: strong
    });
  };
}
function makePlayerHapticReset(player) {
  return async function hapticReset() {
    const gamepad = navigator.getGamepads()[player];
    if (!gamepad)
      return warnDisconnected(player);
    if (!gamepad.vibrationActuator)
      return warnUnsupported(player);
    return await gamepad.vibrationActuator.reset();
  };
}

// src/gamepad.ts
function createStickMotion(player, stick, getAxes, deadzone) {
  function stickMotion(params) {
    const { activate = 0.75, deactivate = 0.25, diagonals = false } = params || {};
    const { rawX, rawY } = getAxes();
    if (diagonals) {
      return motion({
        rawX,
        rawY,
        activationThreshold: activate,
        deactivationThreshold: deactivate,
        deadzone,
        player,
        stick,
        diagonals: true
      });
    } else {
      return motion({
        rawX,
        rawY,
        activationThreshold: activate,
        deactivationThreshold: deactivate,
        deadzone,
        player,
        stick,
        diagonals: false
      });
    }
  }
  return stickMotion;
}
function makeAxes(player) {
  const deadzone = {
    joystick: 0.25,
    trigger: 0.05
  };
  const readAxis2 = (axis) => readAxis(axis, player) ?? 0;
  return {
    readRawAxisValue: readAxis2,
    leftStick: {
      deadzone: deadzone.joystick,
      get x() {
        return normalizeAxisValue(this.rawX, this.deadzone);
      },
      get y() {
        return normalizeAxisValue(this.rawY, this.deadzone);
      },
      get rawX() {
        return readAxis2(Axis.LeftStickX);
      },
      get rawY() {
        return readAxis2(Axis.LeftStickY);
      },
      get magnitude() {
        return clamp(magnitude(this.rawX, this.rawY), 0, 1);
      },
      get angle() {
        return angle(this.rawX, this.rawY);
      },
      get angleDegrees() {
        return radiansToDegrees(this.angle);
      },
      get cardinal() {
        return degreesToCardinalDirection(this.angleDegrees);
      },
      get snap4() {
        return snap4(this.rawX, this.rawY, this.deadzone);
      },
      get snap8() {
        return snap8(this.rawX, this.rawY, this.deadzone);
      },
      motion: createStickMotion(player, Side.Left, () => ({ rawX: readAxis2(Axis.LeftStickX), rawY: readAxis2(Axis.LeftStickY) }), deadzone.joystick)
    },
    rightStick: {
      deadzone: deadzone.joystick,
      get x() {
        return normalizeAxisValue(this.rawX, this.deadzone);
      },
      get y() {
        return normalizeAxisValue(this.rawY, this.deadzone);
      },
      get rawX() {
        return readAxis2(Axis.RightStickX);
      },
      get rawY() {
        return readAxis2(Axis.RightStickY);
      },
      get magnitude() {
        return clamp(magnitude(this.rawX, this.rawY), 0, 1);
      },
      get angle() {
        return angle(this.rawX, this.rawY);
      },
      get angleDegrees() {
        return radiansToDegrees(this.angle);
      },
      get cardinal() {
        return degreesToCardinalDirection(this.angleDegrees);
      },
      get snap4() {
        return snap4(this.rawX, this.rawY, this.deadzone);
      },
      get snap8() {
        return snap8(this.rawX, this.rawY, this.deadzone);
      },
      motion: createStickMotion(player, Side.Right, () => ({ rawX: readAxis2(Axis.RightStickX), rawY: readAxis2(Axis.RightStickY) }), deadzone.joystick)
    },
    leftTrigger: {
      deadzone: deadzone.trigger,
      get value() {
        return normalizeAxisValue(this.rawValue, this.deadzone);
      },
      get rawValue() {
        return readAxis2(Axis.LeftTrigger);
      },
      rumble(duration = 30, intensity = HapticIntensity.Balanced) {
        if (typeof intensity === "number") {
          return leftTriggerHaptic({ player, intensity, duration });
        }
        return leftTriggerHaptic({
          player,
          intensity: mapHapticIntensityToTriggerStrength[intensity],
          duration
        });
      }
    },
    rightTrigger: {
      deadzone: deadzone.trigger,
      get value() {
        return normalizeAxisValue(this.rawValue, this.deadzone);
      },
      get rawValue() {
        return readAxis2(Axis.RightTrigger);
      },
      rumble(duration = 30, intensity = HapticIntensity.Balanced) {
        if (typeof intensity === "number") {
          return rightTriggerHaptic({ player, intensity, duration });
        }
        return rightTriggerHaptic({
          player,
          intensity: mapHapticIntensityToTriggerStrength[intensity],
          duration
        });
      }
    }
  };
}
function makePlayer(player) {
  const value = {
    get index() {
      return player;
    },
    get gamepad() {
      return navigator.getGamepads()[player] ?? null;
    },
    get gamepadInfo() {
      return getGamepadInfo(player);
    },
    get isConnected() {
      return !!navigator.getGamepads()[player]?.connected;
    },
    get justConnected() {
      return gamepadJustConnected(player);
    },
    get justDisconnected() {
      return gamepadJustDisconnected(player);
    },
    isButtonDown(button) {
      return isButtonPressed(button, player);
    },
    wasButtonDown(button) {
      return wasButtonPressed(button, player);
    },
    get buttonsDown() {
      const buttonsDown = new Set;
      const gamepad = this.gamepad;
      if (!gamepad)
        return buttonsDown;
      gamepad.buttons.forEach((button, buttonIndex) => {
        if (button.pressed) {
          buttonsDown.add(buttonIndex);
        }
      });
      return buttonsDown;
    },
    get previousButtonsDown() {
      return previousGamepadButtonsDown[player];
    },
    buttonJustPressed(button) {
      if (resume.player === player) {
        const now = performance.now();
        const cooldown = 1000;
        if (now - resume.time < cooldown && button === Button.South) {
          return false;
        }
      }
      return gamepadButtonJustPressed(button, player);
    },
    buttonJustReleased(button) {
      if (resume.time !== 0 && button === Button.South)
        resume.time = 0;
      return gamepadButtonJustReleased(button, player);
    },
    buttonHeld(button) {
      return gamepadButtonHeld(button, player);
    },
    ...makeAxes(player),
    haptic: makePlayerHaptic(player),
    hapticReset: makePlayerHapticReset(player),
    rumble(duration = 30, intensity = HapticIntensity.Balanced) {
      return this.haptic({
        duration,
        strong: intensity === HapticIntensity.Light ? 0 : intensity === HapticIntensity.Balanced ? 0 : 1,
        weak: intensity === HapticIntensity.Light ? 0.02 : intensity === HapticIntensity.Balanced ? 0.2 : 0.5
      });
    }
  };
  return value;
}
var gamepads = {
  get playerCount() {
    return gamepads.connectedPlayers.list.length;
  },
  anyPlayer: {
    isButtonDown(button) {
      return gamepads.connectedPlayers.list.some((player) => player.isButtonDown(button));
    },
    wasButtonDown(button) {
      return gamepads.connectedPlayers.list.some((player) => player.wasButtonDown(button));
    },
    buttonJustPressed(button) {
      return gamepads.connectedPlayers.list.some((player) => {
        if (player.index === resume.player) {
          const now = performance.now();
          const cooldown = 1000;
          if (now - resume.time < cooldown && button === Button.South) {
            return false;
          }
        }
        return player.buttonJustPressed(button);
      });
    },
    buttonJustReleased(button) {
      return gamepads.connectedPlayers.list.some((player) => player.buttonJustReleased(button));
    },
    buttonHeld(button) {
      return gamepads.connectedPlayers.list.some((player) => player.buttonHeld(button));
    }
  },
  connectedPlayers: {
    get list() {
      return ["p1", "p2", "p3", "p4"].map((player) => gamepads[player]).filter((player) => !!player.gamepad?.connected);
    },
    isButtonDown(button) {
      return gamepads.connectedPlayers.list.every((player) => player.isButtonDown(button));
    },
    buttonHeld(button) {
      return gamepads.connectedPlayers.list.every((player) => player.buttonHeld(button));
    },
    haptic(args) {
      return gamepads.connectedPlayers.list.forEach((player) => player.haptic(args));
    },
    rumble(duration = 30, intensity = HapticIntensity.Balanced) {
      return this.haptic({
        duration,
        strong: intensity === HapticIntensity.Light ? 0 : intensity === HapticIntensity.Balanced ? 0 : 1,
        weak: intensity === HapticIntensity.Light ? 0.02 : intensity === HapticIntensity.Balanced ? 0.2 : 0.5
      });
    },
    leftTrigger: {
      rumble(duration = 30, intensity = HapticIntensity.Balanced) {
        return gamepads.connectedPlayers.list.forEach((player) => player.leftTrigger.rumble(duration, intensity));
      }
    },
    rightTrigger: {
      rumble(duration = 30, intensity = HapticIntensity.Balanced) {
        return gamepads.connectedPlayers.list.forEach((player) => player.rightTrigger.rumble(duration, intensity));
      }
    }
  },
  players: {
    get list() {
      return [gamepads.p1, gamepads.p2, gamepads.p3, gamepads.p4];
    }
  },
  p1: makePlayer(Player.P1),
  p2: makePlayer(Player.P2),
  p3: makePlayer(Player.P3),
  p4: makePlayer(Player.P4),
  get singlePlayer() {
    return this.players.list[lastActiveGamepad];
  },
  clearInputs
};

// src/platform.ts
var colors = {
  [Player.P1]: "oklch(.746 .16 232.661)",
  [Player.P2]: "oklch(.712 .194 13.428)",
  [Player.P3]: "oklch(.852 .199 91.936)",
  [Player.P4]: "oklch(.765 .177 163.223)"
};
var customPlayerNames = {
  [Player.P1]: defaultPlayerNames[Player.P1],
  [Player.P2]: defaultPlayerNames[Player.P2],
  [Player.P3]: defaultPlayerNames[Player.P3],
  [Player.P4]: defaultPlayerNames[Player.P4]
};
var isPaused = false;
var isListening = false;
var abortController = new AbortController;
var spud = {
  p1: {
    get name() {
      return customPlayerNames[Player.P1] || defaultPlayerNames[Player.P1];
    },
    get color() {
      return colors[Player.P1];
    }
  },
  p2: {
    get name() {
      return customPlayerNames[Player.P2] || defaultPlayerNames[Player.P2];
    },
    get color() {
      return colors[Player.P2];
    }
  },
  p3: {
    get name() {
      return customPlayerNames[Player.P3] || defaultPlayerNames[Player.P3];
    },
    get color() {
      return colors[Player.P3];
    }
  },
  p4: {
    get name() {
      return customPlayerNames[Player.P4] || defaultPlayerNames[Player.P4];
    },
    get color() {
      return colors[Player.P4];
    }
  },
  singlePlayer: {
    get name() {
      const player = gamepads.singlePlayer.index;
      return customPlayerNames[player] || defaultPlayerNames[player];
    },
    get color() {
      const player = gamepads.singlePlayer.index;
      return colors[player];
    }
  },
  get players() {
    return [spud.p1, spud.p2, spud.p3, spud.p4];
  },
  get connectedPlayers() {
    return gamepads.connectedPlayers.list.map((player) => ({
      get name() {
        return customPlayerNames[player.index] || defaultPlayerNames[player.index];
      },
      get color() {
        return colors[player.index];
      }
    }));
  },
  get isDemoMode() {
    const params = new URLSearchParams(window.location.search);
    return params.get("mode") === "demo";
  },
  get isPaused() {
    return isPaused;
  },
  pauseGame() {
    if (window.parent !== window) {
      window.parent.postMessage("spud:pause");
    }
    isPaused = true;
  },
  listen() {
    if (isListening) {
      console.warn("spud api already listening for message and visibilitychange events. skipping reinitialization.");
      return;
    }
    isListening = true;
    window.addEventListener("visibilitychange", () => {
      if (document.hidden && window.parent !== window && !isPaused && !this.isDemoMode) {
        window.parent.postMessage("spud:pause");
        isPaused = true;
      }
    }, { signal: abortController.signal });
    window.addEventListener("message", (event) => {
      try {
        const data = JSON.parse(event.data);
        if (typeof data !== "object" || data === null)
          return;
        if (!("type" in data))
          return;
        switch (data.type) {
          case "pause": {
            isPaused = true;
            return;
          }
          case "resume": {
            resume.time = performance.now();
            const possiblePlayers = [...playerIndexes, AnyPlayer];
            if (!("player" in data))
              return;
            const player = data.player;
            resume.player = includes(possiblePlayers, player) ? player : AnyPlayer;
            isPaused = false;
            return;
          }
          case "playerinfo": {
            if (!("players" in data))
              return;
            const players = data.players;
            if (!Array.isArray(players))
              return;
            playerIndexes.forEach((playerIndex) => {
              const name = players[playerIndex]?.name;
              if (typeof name !== "string")
                return;
              if (name.length === 0 || name.length > 3)
                return;
              customPlayerNames[playerIndex] = name;
            });
            break;
          }
          default: {
            console.warn("Unknown message type", data.type);
            break;
          }
        }
      } catch (err) {
        console.warn("Unable to parse postMessage data");
        console.error(err);
      }
    }, { signal: abortController.signal });
  },
  disconnect() {
    abortController.abort();
    abortController = new AbortController;
    isListening = false;
  },
  get isListening() {
    return isListening;
  },
  colors
};
function createPauseEventMessage() {
  return { type: "pause", player: AnyPlayer };
}
function createResumeEventMessage(player) {
  return { type: "resume", player };
}
function createPlayerInfoMessage({
  players
}) {
  return { type: "playerinfo", players };
}
// src/easings.ts
var exports_easings = {};
__export(exports_easings, {
  linear: () => linear,
  easeOutSine: () => easeOutSine,
  easeOutQuint: () => easeOutQuint,
  easeOutQuart: () => easeOutQuart,
  easeOutQuad: () => easeOutQuad,
  easeOutExpo: () => easeOutExpo,
  easeOutElastic: () => easeOutElastic,
  easeOutCubic: () => easeOutCubic,
  easeOutCirc: () => easeOutCirc,
  easeOutBounce: () => easeOutBounce,
  easeOutBack: () => easeOutBack,
  easeInSine: () => easeInSine,
  easeInQuint: () => easeInQuint,
  easeInQuart: () => easeInQuart,
  easeInQuad: () => easeInQuad,
  easeInOutSine: () => easeInOutSine,
  easeInOutQuint: () => easeInOutQuint,
  easeInOutQuart: () => easeInOutQuart,
  easeInOutQuad: () => easeInOutQuad,
  easeInOutExpo: () => easeInOutExpo,
  easeInOutElastic: () => easeInOutElastic,
  easeInOutCubic: () => easeInOutCubic,
  easeInOutCirc: () => easeInOutCirc,
  easeInOutBounce: () => easeInOutBounce,
  easeInOutBack: () => easeInOutBack,
  easeInExpo: () => easeInExpo,
  easeInElastic: () => easeInElastic,
  easeInCubic: () => easeInCubic,
  easeInCirc: () => easeInCirc,
  easeInBounce: () => easeInBounce,
  easeInBack: () => easeInBack
});
var pow = Math.pow;
var sqrt = Math.sqrt;
var sin = Math.sin;
var cos = Math.cos;
var pi = Math.PI;
var c1 = 1.70158;
var c2 = c1 * 1.525;
var c3 = c1 + 1;
var c4 = 2 * pi / 3;
var c5 = 2 * pi / 4.5;
function linear(x) {
  return x;
}
function easeInQuad(x) {
  return x * x;
}
function easeOutQuad(x) {
  return 1 - (1 - x) * (1 - x);
}
function easeInOutQuad(x) {
  return x < 0.5 ? 2 * x * x : 1 - pow(-2 * x + 2, 2) / 2;
}
function easeInCubic(x) {
  return x * x * x;
}
function easeOutCubic(x) {
  return 1 - pow(1 - x, 3);
}
function easeInOutCubic(x) {
  return x < 0.5 ? 4 * x * x * x : 1 - pow(-2 * x + 2, 3) / 2;
}
function easeInQuart(x) {
  return x * x * x * x;
}
function easeOutQuart(x) {
  return 1 - pow(1 - x, 4);
}
function easeInOutQuart(x) {
  return x < 0.5 ? 8 * x * x * x * x : 1 - pow(-2 * x + 2, 4) / 2;
}
function easeInQuint(x) {
  return x * x * x * x * x;
}
function easeOutQuint(x) {
  return 1 - pow(1 - x, 5);
}
function easeInOutQuint(x) {
  return x < 0.5 ? 16 * x * x * x * x * x : 1 - pow(-2 * x + 2, 5) / 2;
}
function easeInSine(x) {
  return 1 - cos(x * pi / 2);
}
function easeOutSine(x) {
  return sin(x * pi / 2);
}
function easeInOutSine(x) {
  return -(cos(pi * x) - 1) / 2;
}
function easeInExpo(x) {
  return x === 0 ? 0 : pow(2, 10 * x - 10);
}
function easeOutExpo(x) {
  return x === 1 ? 1 : 1 - pow(2, -10 * x);
}
function easeInOutExpo(x) {
  return x === 0 ? 0 : x === 1 ? 1 : x < 0.5 ? pow(2, 20 * x - 10) / 2 : (2 - pow(2, -20 * x + 10)) / 2;
}
function easeInCirc(x) {
  return 1 - sqrt(1 - pow(x, 2));
}
function easeOutCirc(x) {
  return sqrt(1 - pow(x - 1, 2));
}
function easeInOutCirc(x) {
  return x < 0.5 ? (1 - sqrt(1 - pow(2 * x, 2))) / 2 : (sqrt(1 - pow(-2 * x + 2, 2)) + 1) / 2;
}
function easeInBack(x) {
  return c3 * x * x * x - c1 * x * x;
}
function easeOutBack(x) {
  return 1 + c3 * pow(x - 1, 3) + c1 * pow(x - 1, 2);
}
function easeInOutBack(x) {
  return x < 0.5 ? pow(2 * x, 2) * ((c2 + 1) * 2 * x - c2) / 2 : (pow(2 * x - 2, 2) * ((c2 + 1) * (x * 2 - 2) + c2) + 2) / 2;
}
function easeInElastic(x) {
  return x === 0 ? 0 : x === 1 ? 1 : -pow(2, 10 * x - 10) * sin((x * 10 - 10.75) * c4);
}
function easeOutElastic(x) {
  return x === 0 ? 0 : x === 1 ? 1 : pow(2, -10 * x) * sin((x * 10 - 0.75) * c4) + 1;
}
function easeInOutElastic(x) {
  return x === 0 ? 0 : x === 1 ? 1 : x < 0.5 ? -(pow(2, 20 * x - 10) * sin((20 * x - 11.125) * c5)) / 2 : pow(2, -20 * x + 10) * sin((20 * x - 11.125) * c5) / 2 + 1;
}
function easeInBounce(x) {
  return 1 - easeOutBounce(1 - x);
}
function easeOutBounce(x) {
  const n1 = 7.5625;
  const d1 = 2.75;
  if (x < 1 / d1) {
    return n1 * x * x;
  } else if (x < 2 / d1) {
    const t = x - 1.5 / d1;
    return n1 * t * t + 0.75;
  } else if (x < 2.5 / d1) {
    const t = x - 2.25 / d1;
    return n1 * t * t + 0.9375;
  } else {
    const t = x - 2.625 / d1;
    return n1 * t * t + 0.984375;
  }
}
function easeInOutBounce(x) {
  return x < 0.5 ? (1 - easeOutBounce(1 - 2 * x)) / 2 : (1 + easeOutBounce(2 * x - 1)) / 2;
}
// src/audio.ts
var exports_audio = {};
__export(exports_audio, {
  retrieveOrCreateGlobalAudio: () => retrieveOrCreateGlobalAudio,
  createSounds: () => createSounds,
  createMusic: () => createMusic
});
var cachedGlobalAudio = null;
function retrieveOrCreateGlobalAudio() {
  const { audioContext, masterGain, sfxGain, musicGain } = cachedGlobalAudio || (() => {
    const audioContext2 = new AudioContext({ latencyHint: "interactive" });
    const masterGain2 = audioContext2.createGain();
    const sfxGain2 = audioContext2.createGain();
    const musicGain2 = audioContext2.createGain();
    sfxGain2.connect(masterGain2);
    musicGain2.connect(masterGain2);
    masterGain2.connect(audioContext2.destination);
    cachedGlobalAudio = { audioContext: audioContext2, masterGain: masterGain2, sfxGain: sfxGain2, musicGain: musicGain2 };
    resumeAudioContextOnInteraction(audioContext2);
    return { audioContext: audioContext2, masterGain: masterGain2, sfxGain: sfxGain2, musicGain: musicGain2 };
  })();
  return { audioContext, masterGain, sfxGain, musicGain };
}
function resumeAudioContextOnInteraction(audioContext) {
  const abortController2 = new AbortController;
  const interactionEvents = ["click", "contextmenu", "auxclick", "dblclick", "mousedown", "mouseup", "pointerup", "touchend", "keydown", "keyup"];
  const resume2 = () => {
    if (audioContext.state === "suspended")
      audioContext.resume();
    abortController2.abort();
  };
  if (!spud.isDemoMode) {
    interactionEvents.forEach((eventName) => {
      document.addEventListener(eventName, resume2, {
        passive: true,
        signal: abortController2.signal
      });
    });
  }
  return () => abortController2.abort();
}
function createBufferCache(audioContext) {
  const cache = {};
  async function bufferAudio(url) {
    const cached = cache[url];
    if (cached)
      return cached;
    const buffer = await fetch(url).then((response) => response.arrayBuffer()).then((arrayBuffer) => audioContext.decodeAudioData(arrayBuffer));
    cache[url] = buffer;
    return buffer;
  }
  return { cache, bufferAudio };
}
function createSounds(sources) {
  const { audioContext, sfxGain, masterGain, musicGain } = retrieveOrCreateGlobalAudio();
  const { cache, bufferAudio } = createBufferCache(audioContext);
  const handles = Object.fromEntries(Object.entries(sources).map(([key, { url, volume: defaultVolume = 0.5 }]) => {
    const soundGain = audioContext.createGain();
    soundGain.gain.value = defaultVolume;
    soundGain.connect(sfxGain);
    const handle = {
      buffer: null,
      isBuffered: false,
      play(options) {
        if (spud.isDemoMode)
          return { success: false, reason: "spud.isDemoMode" };
        const { volume = 1, detune = 0, playbackRate = 1 } = options ?? {};
        const buffer = cache[url];
        if (!buffer)
          return { success: false, reason: "buffering" };
        const source = audioContext.createBufferSource();
        source.detune.value = detune;
        source.playbackRate.value = playbackRate;
        source.buffer = buffer;
        const playGain = audioContext.createGain();
        playGain.gain.value = volume;
        source.connect(playGain);
        playGain.connect(soundGain);
        source.start();
        return { success: true, gain: playGain, source };
      }
    };
    bufferAudio(url).then((buffer) => {
      const handle2 = handles[key];
      assert(handle2);
      handle2.buffer = buffer;
      handle2.isBuffered = true;
    });
    return [key, handle];
  }));
  const get = (name) => handles[name];
  const globalAudio = { audioContext, masterGain, sfxGain, musicGain };
  get.global = globalAudio;
  return get;
}
function createMusic(source) {
  const { audioContext, musicGain, masterGain, sfxGain } = retrieveOrCreateGlobalAudio();
  const globalAudio = { audioContext, masterGain, sfxGain, musicGain };
  const gain = audioContext.createGain();
  gain.gain.value = source.volume ?? 0.5;
  gain.connect(musicGain);
  const element = new Audio(source.url);
  element.preload = "auto";
  element.loop = source.loop ?? false;
  const elementSource = audioContext.createMediaElementSource(element);
  elementSource.connect(gain);
  const handle = {
    async play() {
      if (spud.isDemoMode)
        return { success: false, reason: "spud.isDemoMode" };
      try {
        if (audioContext.state === "suspended")
          await audioContext.resume();
        await element.play();
        return { success: true };
      } catch (error) {
        return { success: false, reason: "playback", error };
      }
    },
    pause() {
      element.pause();
    },
    stop() {
      element.pause();
      element.currentTime = 0;
    },
    fadeTo({ volume, duration = 0.25 }) {
      const now = audioContext.currentTime;
      const safeDuration = Math.max(0, duration);
      gain.gain.cancelScheduledValues(now);
      if (safeDuration === 0) {
        gain.gain.setValueAtTime(volume, now);
        return;
      }
      gain.gain.setValueAtTime(gain.gain.value, now);
      gain.gain.linearRampToValueAtTime(volume, now + safeDuration);
    },
    get volume() {
      return gain.gain.value;
    },
    get currentTime() {
      return element.currentTime;
    },
    get isPaused() {
      return element.paused;
    },
    get duration() {
      return element.duration;
    },
    audioContext,
    gain,
    element,
    global: globalAudio
  };
  return handle;
}
// src/synth.ts
var exports_synth = {};
__export(exports_synth, {
  noteFrequencies: () => noteFrequencies,
  create: () => create
});
var baseNoteFrequencies = {
  C: [16.35, 32.7, 65.41, 130.81, 261.63, 523.25, 1046.5, 2093, 4186.01],
  Db: [17.32, 34.65, 69.3, 138.59, 277.18, 554.37, 1108.73, 2217.46, 4434.92],
  D: [18.35, 36.71, 73.42, 146.83, 293.66, 587.33, 1174.66, 2349.32, 4698.64],
  Eb: [19.45, 38.89, 77.78, 155.56, 311.13, 622.25, 1244.51, 2489.02, 4978.03],
  E: [20.6, 41.2, 82.41, 164.81, 329.63, 659.26, 1318.51, 2637.02],
  F: [21.83, 43.65, 87.31, 174.61, 349.23, 698.46, 1396.91, 2793.83],
  Gb: [23.12, 46.25, 92.5, 185, 369.99, 739.99, 1479.98, 2959.96],
  G: [24.5, 49, 98, 196, 392, 783.99, 1567.98, 3135.96],
  Ab: [25.96, 51.91, 103.83, 207.65, 415.3, 830.61, 1661.22, 3322.44],
  A: [27.5, 55, 110, 220, 440, 880, 1760, 3520],
  Bb: [29.14, 58.27, 116.54, 233.08, 466.16, 932.33, 1864.66, 3729.31],
  B: [30.87, 61.74, 123.47, 246.94, 493.88, 987.77, 1975.53, 3951.07]
};
var noteFrequencies = {
  ...baseNoteFrequencies,
  "C#": baseNoteFrequencies.Db,
  "D#": baseNoteFrequencies.Eb,
  "F#": baseNoteFrequencies.Gb,
  "G#": baseNoteFrequencies.Ab,
  "A#": baseNoteFrequencies.Bb
};
function create({ wave = "sine", _defaultFrequency = 440 }) {
  const { audioContext, masterGain } = retrieveOrCreateGlobalAudio();
  return {
    audioContext,
    masterGain,
    get currentTime() {
      return audioContext.currentTime;
    },
    createConnectedOscillator() {
      const oscillator = audioContext.createOscillator();
      const gainNode = audioContext.createGain();
      gainNode.connect(masterGain);
      oscillator.connect(gainNode);
      oscillator.onended = () => {
        oscillator.disconnect();
        gainNode.disconnect();
      };
      return { oscillator, gainNode };
    },
    play({
      attackTime = 0,
      decayTime = 0,
      sustainTime = 1,
      releaseTime = 0,
      sustainLevel = 0.5,
      peakLevel = 0.5,
      frequency = noteFrequencies.A[4],
      customize = () => {}
    }) {
      const { gainNode, oscillator } = this.createConnectedOscillator();
      const start = audioContext.currentTime;
      const attackEnd = start + attackTime;
      const decayEnd = attackEnd + decayTime;
      const sustainEnd = decayEnd + sustainTime;
      const releaseEnd = sustainEnd + releaseTime;
      const { gain } = gainNode;
      oscillator.frequency.setValueAtTime(frequency, start);
      oscillator.type = wave;
      gain.cancelScheduledValues(start);
      if (attackTime !== 0) {
        gain.setValueAtTime(Number.EPSILON, start);
        gain.exponentialRampToValueAtTime(peakLevel, attackEnd);
      }
      if (decayTime !== 0) {
        gain.exponentialRampToValueAtTime(Math.max(Number.EPSILON, sustainLevel), decayEnd);
      }
      gain.setValueAtTime(sustainLevel, decayEnd);
      gain.setValueAtTime(sustainLevel, sustainEnd);
      if (releaseTime === 0) {
        gain.setValueAtTime(0, sustainEnd);
      } else {
        gain.exponentialRampToValueAtTime(Number.EPSILON, releaseEnd);
      }
      customize({ oscillator, gain: gainNode, audioContext, masterGain });
      oscillator.start(start);
      oscillator.stop(releaseEnd + 0.01);
    }
  };
}

// src/index.ts
var internal = {
  createPauseEventMessage,
  createResumeEventMessage,
  createPlayerInfoMessage
};
export {
  exports_synth as synth,
  spud,
  playerProperties,
  playerIndexes,
  internal,
  identifyGamepad,
  haptic,
  gamepads,
  exports_gamepad_util as gamepadUtil,
  exports_easings as easing,
  defaultPlayerNames,
  buttons,
  axes,
  exports_audio as audio,
  Side,
  PlayerProperty,
  Player,
  Neutral,
  Motion,
  HapticIntensity,
  GamepadModel,
  GamepadBaseLayout,
  CardinalDirection,
  Button,
  Axis,
  AnyPlayer,
  AllPlayers
};
